### PR TITLE
Change names and add documentation for new pipeline

### PR DESCRIPTION
**Pull request by Aflynn50**
Change type/variable names and add documentation for the new SQLair pipeline. A `doc.go` file is added to the `expr` package to detail how the pipeline should look. The names of the types in the pipeline are changed to clarify they purpose.

The full planned changes are described in the `doc.go` file added here. The upcoming changes In short are: the SQL method will be on `PrimedQuery` rather than `TypedExpr` and the `TypedExpr` will be changed to retain the shape of the AST from the `ParsedExpr`.

### CODE REVIEW

File: internal/expr/doc.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+# Parsing stage
+
+The parsing stage takes a SQLair query string and parses it into an AST. The
+parser only processes information already encoded in the syntax of the SQLair
+expressions.
+
+# Type Binding stage
+
+The Type Binding stage binds concrete Go types to the type names in the SQLair
+query expressions. Type samples are provided by the user. These are used to
```
Comment (letFunny) [inline internal/expr/doc.go]: Should we drop the `query` from `query expressions`? I see above we are saying `SQLair expressions` which seems more natural.

File: internal/expr/doc.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+validate the SQLair query expressions, and to generate an intermediate
```
Comment (letFunny) [inline internal/expr/doc.go]: Ditto.

File: internal/expr/doc.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (letFunny) [inline internal/expr/doc.go:1]: I love the idea of including the structure of the pipeline in the code itself. I was a bit reluctant but I can see how newcomers and contributors would find it very useful.

I guess the idea is to include the docs as part of this PR and update the implementation to match it in subsequent ones, is that right?

File: internal/expr/parser.go [hunk 4]
```diff
@@ -6,15 +6,13 @@ import (
 	"strings"
 )
 
-// A QueryPart represents a section of a parsed SQL statement, which forms
-// a complete query when processed together with its surrounding parts, in
-// their correct order.
-type queryPart interface {
-	// String returns the part's representation for debugging purposes.
+// expr represents a section of a parsed SQL statement.
```
Comment (letFunny) [inline internal/expr/parser.go]: I would prefer avoiding the usage of `statement` which is already very overloaded in SQLair. Additionally, what do you think of changing the description to state it is conceptually part of an AST, not just a mere section of the text in the query. Maybe something along the lines of:
```
// expr are the nodes of the query's AST
```
or
```
// expr are the constructs that constitute a parsed SQL query
```

File: internal/expr/parser.go [hunk 5]
```diff
@@ -40,54 +38,54 @@ func (ca columnAccessor) String() string {
...
-	prevPart int
-	// partStart is the value of pos just before we started parsing the part
-	// under pos. We maintain partStart >= prevPart.
-	partStart int
-	parts     []queryPart
+	// prevExpr is the value of pos when we last finished parsing a expr.
+	prevExpr int
+	// exprStart is the value of pos just before we started parsing the expr
+	// under pos. We maintain exprStart >= prevExpr.
+	exprStart int
```
Comment (letFunny) [inline internal/expr/parser.go]: [Nit]: Now that we are changing the names, either as part of this PR or another one can we change the names to `prevExprPos`, `exprStartPos`, or even `prevExprEnd` and `currentExprStart`? We can discuss them offline but something that is self-explanatory would be perfect.

File: internal/expr/parser.go [hunk 6]
```diff
@@ -257,37 +249,36 @@ func (p *Parser) skipComment() bool {
 	return false
 }
 
-// Parse takes an input string and parses the input and output parts. It returns
-// a pointer to a ParsedExpr.
-func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
+// Parse takes an SQLair query string and returns a ParsedExpr.
+func (p *Parser) Parse(query string) (pe *ParsedExpr, err error) {
```
Comment (letFunny) [inline internal/expr/parser.go]: [Nit]: I am not sure we should rename the argument to `query` here. `query` means something different in various sites in the code base while `input` has a concrete meaning. And then in other PRs (function parsing) we are using the parser to parse things that are not queries which is why I think `input` is better.

File: internal/expr/query.go [hunk 7]
```diff
@@ -87,19 +87,22 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
...
 var scannerInterface = reflect.TypeOf((*sql.Scanner)(nil)).Elem()
 
-// ScanArgs returns list of pointers to the struct fields that are listed in qe.outputs.
-// All the structs and maps mentioned in the query must be in outputArgs.
-func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) (scanArgs []any, onSuccess func(), err error) {
+// ScanArgs takes the result column names and the SQLair output arguments. It
+// returns list of pointers to be used as arguments for rows.Scan. They point to
+// locations within the SQLair output arguments. The onSuccess function should
+// be run once the pointers have been scanned to populate output arguments that
+// cannot be written to directly (such as map values).
```
Comment (letFunny) [inline internal/expr/query.go]: I think this comment can be improved a little if we reword it to avoid saying "arguments" for two different things, and if we avoid talking about the implementation (i.e. all we care about is that `onSuccess` needs to be ran after rows.Scan, we do not care why in this comment).

What do you think of the following?
```
// ScanArgs produces a list of pointers to be passed to rows.Scan. After a
// successful call, the onSuccess function must be invoked. The end result is
// that the outputArgs will be populated with the results from the query.
// All the structs/maps/slices mentioned in the query must be in outputArgs.
```
plus changing `columns` to `columnNames`.

File: sqlair.go [hunk 8]
```diff
@@ -50,7 +50,7 @@ var cacheMutex sync.RWMutex
 // It is ready to be run on a SQLair DB.
 type Statement struct {
 	cacheID stmtID
-	pe      *expr.PreparedExpr
+	te      *expr.TypedExpr
```
Comment (letFunny) [inline sqlair.go]: What do you think about including a comments here saying we are caching the `TypeExpr` to be reused for all queries? It used to be that the `sqlair.Statement` had all the values necessary so it was self-explanatory but that is no longer that.

File: internal/expr/query.go [hunk 9]
```diff
@@ -8,36 +8,36 @@ import (
...
+	return len(qv.outputs) > 0
 }
 
-type QueryExpr struct {
+// QueryValues contains all concrete values needed to run a SQLair query on a
+// database.
+type QueryValues struct {
 	sql     string
-	args    []any
+	params  []any
```
Comment (letFunny) [inline internal/expr/query.go]: We have probably talked about this already, but why is this named parameters if the types contained are named `NamedArg` implying arguments (we might have to change the name in the reference spec if we change it here)?

**Review by letFunny — COMMENTED** [review]
This PR was very helpful for seeing the actual usage of the new names in the code, thanks! I already liked the new names but now seeing them in the actual code, I can see how well they fit.

I left a couple of suggestions mostly regarding small wording changes.

File: internal/expr/doc.go (hunk 1 above)
Comment (Aflynn50) [reply internal/expr/doc.go]: True, changed to just "SQLair expressions".

File: internal/expr/doc.go (hunk 2 above)
Comment (Aflynn50) [reply internal/expr/doc.go]: Ditto.

File: internal/expr/doc.go (hunk 3 above)
Comment (Aflynn50) [reply internal/expr/doc.go:1]: That's right, I've updated the PR comment to reflect this.

File: internal/expr/parser.go (hunk 4 above)
Comment (Aflynn50) [reply internal/expr/parser.go]: Changed to: `expr represents a parsed node of the SQLair query's AST.`

File: internal/expr/parser.go (hunk 5 above)
Comment (Aflynn50) [reply internal/expr/parser.go]: I like those suggested names, they are a lot clearer than the current names, I'll change them.

File: internal/expr/parser.go (hunk 6 above)
Comment (Aflynn50) [reply internal/expr/parser.go]: I think query makes sense in regard to this being a public method, we want to say what the argument is to the user, not to us. But since this is in the internal package it makes sense to change it back to input.

File: internal/expr/query.go (hunk 7 above)
Comment (Aflynn50) [reply internal/expr/query.go]: Fair enough, the previous comment was a bit rambly, I've changed it to a more concise version of your suggestion.

File: sqlair.go (hunk 8 above)
Comment (Aflynn50) [reply sqlair.go]: I've added the comment:
```
// te is the type checked SQLair query. It is used to generate the required
// query values when the Statement is run on a database.
```

File: internal/expr/query.go (hunk 9 above)
Comment (Aflynn50) [reply internal/expr/query.go]: In the spec we are using "Input arguments" to refer to the structs and maps from the user, and "query parameters" to refer to what is passed to the database. I've changed this name to try and disambiguate the two in the code.

I've added a comment here (and one for `outputs` too)

File: sqlair.go (hunk 8 above)
Comment (letFunny) [reply sqlair.go]: Make it `type bounded` to be consistent with the terminology. I like the comments but I think we can better convey that we are only caching the tree. We really don't need this in the statement, we have all the information in the .Query and .Get args as well. And `it is used to generate` seems to imply that we are using this alone to generate the values when in reality we have to wait for the .Query step.

Comment (Aflynn50) [reply sqlair.go]: I've changed it from type checked to type bound.

I'm avoiding using the word caching here to talk about the information stored in the type bound expression because the other field in this struct is called `cacheID` and refers to something else. I wouldn't say that we are simply caching the tree here. The SQL string only appears in the call to `Prepare` and we have added the type information, some of which only appears at the `.Get` stage and is needed at the `.Query` stage.

I've changed the comment to:
```
te is the type bound SQLair query. It contains information used to
generate query values from the input arguments when the Statement is 
run on a database.
```

Comment (letFunny) [reply sqlair.go]: I think the new comment you suggested is perfect. And I agree we should not talk about caching in two different contexts which might get muddled.

File: internal/expr/parser.go [hunk 10]
```diff
@@ -6,15 +6,13 @@ import (
...
-type queryPart interface {
-	// String returns the part's representation for debugging purposes.
+// expression represents a parsed node of the SQLair query's AST.
+type expression interface {
+	// String returns a text representation for debugging and testing purposes.
 	String() string
 
 	// marker method
-	part()
+	expression()
```
Comment (manadart) [inline internal/expr/parser.go]: Keep this one as `expr` to make it distinct from the type name. I hope we can obviate the meed for it in future anyway.

File: internal/expr/query.go [hunk 11]
```diff
@@ -8,36 +8,38 @@ import (
...
-// substitute the input placeholders in the prepared expression. These placeholders use
-// the syntax "$Person.fullname", where Person would be a type such as:
-//
-//	type Person struct {
-//	        Name string `db:"fullname"`
-//	}
-func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
+// BindInputs takes the SQLair input arguments and returns the PrimedQuery ready
+// for use with the database.
+func (te *TypedExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
```
Comment (manadart) [inline internal/expr/query.go]: Move this to be co-located with its type declaration in _typebind.go_

File: internal/expr/query.go [hunk 12]
```diff
@@ -69,9 +71,9 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
 		}
 	}
 
-	// Query parameteres.
-	qargs := []any{}
-	for i, typeMember := range pe.inputs {
+	// Query parameters.
```
Comment (manadart) [inline internal/expr/query.go]: Not for this patch, but this is an example of what I mentioned in our meeting. If you are using comments to label the steps of a large method, just make the steps their own methods.

File: internal/expr/query.go [hunk 13]
```diff
@@ -8,36 +8,38 @@ import (
 	"strings"
 )
 
-func (qe *QueryExpr) QueryArgs() []any {
-	return qe.args
+// PrimedQuery contains all concrete values needed to run a SQLair query on a
+// database.
+type PrimedQuery struct {
```
Comment (manadart) [inline internal/expr/query.go:12]: I had a thought that `PrimedQuery` is not such a great name due to being a member of `sqlair.Query`, but I think it's OK, because it is distinct by being from the `expr` package, from who's point-of-view the work is done.

**Review by manadart — COMMENTED** [review]
Almost there.

File: internal/expr/parser.go (hunk 10 above)
Comment (Aflynn50) [reply internal/expr/parser.go]: Changed

File: internal/expr/query.go (hunk 13 above)
Comment (Aflynn50) [reply internal/expr/query.go:12]: I'm not 100% happy with it either, but it can stay there for now, if we think of a better name we can change it

File: internal/expr/query.go (hunk 11 above)
Comment (Aflynn50) [reply internal/expr/query.go]: Moved to `typebind.go`.

Following this pattern the `func (*ParsedExpr) BindTypes(...)` method in `typebind.go` should really be moved to the parser to be colocated with the declaration of `ParsedExpr`. To me though that move would feel a bit wrong, all the actual "type binding" takes place in that method so it should really be in `bindtypes.go`. I have left it as it is for now.

**Review by manadart — APPROVED** [review]
This is good to land and iterate on.
